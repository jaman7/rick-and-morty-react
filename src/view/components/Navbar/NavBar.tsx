import classNames from 'classnames';
import { NavLink } from 'react-router-dom';

interface MenuItem {
  name: string;
  to: string;
}

interface Props {
  isMobileMenuOpen?: boolean;
  onClose?: () => void;
}

const Navbar: React.FC<Props> = ({ isMobileMenuOpen = false, onClose }) => {
  const menuData: MenuItem[] = [
    {
      name: 'Characters',
      to: '/',
    },
    {
      name: 'Episodes',
      to: '/episodes',
    },
    {
      name: 'Location',
      to: '/location',
    },
  ];

  return (
    <nav
      id="main-navigation"
      className={classNames('navbar', { 'navbar--mobile': isMobileMenuOpen, 'navbar--visible': isMobileMenuOpen })}
      role="navigation"
      aria-label="Main navigation"
    >
      <ul className={classNames('menu')}>
        {menuData?.map((item, i) => (
          <li key={`menu-item-${i}`} onClick={onClose}>
            <NavLink
              className={({ isActive, isPending }) =>
                classNames('item', {
                  pending: isPending,
                  active: isActive,
                })
              }
              to={item.to as string}
            >
              <span>{item.name}</span>
            </NavLink>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default Navbar;
