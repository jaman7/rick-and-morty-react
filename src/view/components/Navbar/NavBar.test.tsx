import { render, screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import Navbar from './NavBar';

const renderNavbar = (initialPath = '/', isMobileMenuOpen = false) => {
  return render(
    <MemoryRouter initialEntries={[initialPath]}>
      <Navbar isMobileMenuOpen={isMobileMenuOpen} />
    </MemoryRouter>
  );
};

describe('Navbar', () => {
  it('renders all navigation links with correct labels and hrefs', () => {
    renderNavbar();

    const characters = screen.getByRole('link', { name: /Characters/i });
    const episodes = screen.getByRole('link', { name: /Episodes/i });
    const location = screen.getByRole('link', { name: /Location/i });

    expect(characters).toHaveAttribute('href', '/');
    expect(episodes).toHaveAttribute('href', '/episodes');
    expect(location).toHaveAttribute('href', '/location');
  });

  it('adds active class when link matches current route', () => {
    renderNavbar('/episodes');

    const episodes = screen.getByRole('link', { name: /Episodes/i });
    expect(episodes.className).toMatch(/active/);
  });

  it('applies mobile menu classes when open', () => {
    const { container } = renderNavbar('/', true);

    const nav = container.querySelector('nav');
    expect(nav?.className).toMatch(/navbar--mobile/);
    expect(nav?.className).toMatch(/navbar--visible/);
  });

  it('matches snapshot', () => {
    const { container } = renderNavbar('/', true);
    expect(container).toMatchSnapshot();
  });
});
