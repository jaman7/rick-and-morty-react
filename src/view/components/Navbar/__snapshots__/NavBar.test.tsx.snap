// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > matches snapshot 1`] = `
<div>
  <nav
    aria-label="Main navigation"
    class="navbar navbar--mobile navbar--visible"
    id="main-navigation"
    role="navigation"
  >
    <ul
      class="menu"
    >
      <li>
        <a
          aria-current="page"
          class="item active"
          data-discover="true"
          href="/"
        >
          <span>
            Characters
          </span>
        </a>
      </li>
      <li>
        <a
          class="item"
          data-discover="true"
          href="/episodes"
        >
          <span>
            Episodes
          </span>
        </a>
      </li>
      <li>
        <a
          class="item"
          data-discover="true"
          href="/location"
        >
          <span>
            Location
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
